ssl/quic/quic_impl.c

```diff
@@ -0,0 +1,187 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/macros.h>
+#include <openssl/objects.h>
+#include "quic_local.h"
+
+int ossl_quic_new(SSL *s)
+{
+    return s->method->ssl_clear(s);
+}
+
+void ossl_quic_free(SSL *s)
+{
+    return;
+}
+
+int ossl_quic_clear(SSL *s)
+{
+    return 1;
+}
+
+int ossl_quic_accept(SSL *s)
+{
+    s->statem.in_init = 0;
+    return 1;
+}
+
+int ossl_quic_connect(SSL *s)
+{
+    s->statem.in_init = 0;
+    return 1;
+}
+
+int ossl_quic_read(SSL *s, void *buf, size_t len, size_t *readbytes)
+{
+    int ret;
+    BIO *rbio = SSL_get_rbio(s);
+
+    if (rbio == NULL)
+        return 0;
+
+    s->rwstate = SSL_READING;
+    ret = BIO_read_ex(rbio, buf, len, readbytes);
+    return ret <= 0 ? -1 : ret;
+}
+
+int ossl_quic_peek(SSL *s, void *buf, size_t len, size_t *readbytes)
+{
+    return 1;
+}
+
+int ossl_quic_write(SSL *s, const void *buf, size_t len, size_t *written)
+{
+    BIO *wbio = SSL_get_wbio(s);
+
+    if (wbio == NULL)
+        return 0;
+
+    s->rwstate = SSL_WRITING;
+    return BIO_write_ex(wbio, buf, len, written);
+}
+
+int ossl_quic_shutdown(SSL *s)
+{
+    return 1;
+}
+
+long ossl_quic_ctrl(SSL *s, int cmd, long larg, void *parg)
+{
+    switch(cmd) {
+    case SSL_CTRL_CHAIN:
+        if (larg)
+            return ssl_cert_set1_chain(s, NULL, (STACK_OF(X509) *)parg);
+        else
+            return ssl_cert_set0_chain(s, NULL, (STACK_OF(X509) *)parg);
+    }
+    return 0;
+}
+
+long ossl_quic_ctx_ctrl(SSL_CTX *ctx, int cmd, long larg, void *parg)
+{
+    switch(cmd) {
+    case SSL_CTRL_CHAIN:
+        if (larg)
+            return ssl_cert_set1_chain(NULL, ctx, (STACK_OF(X509) *)parg);
+        else
+            return ssl_cert_set0_chain(NULL, ctx, (STACK_OF(X509) *)parg);
+
+    case SSL_CTRL_SET_TLSEXT_TICKET_KEYS:
+    case SSL_CTRL_GET_TLSEXT_TICKET_KEYS:
```

Comment:
They are necessary for the ssl_test_new machinery. I did not want to modify ssl_test_new because in the end we IMO want these CTRLS to work with the SSL connection object, don't we?

Reply:
BTW, I suppose I could just return 1 instead of copying the whole implementation, if you prefer this.

Reply:
I think that's probably better. Perhaps with a `TODO(QUIC)` comment.